### PR TITLE
Add support for regexp_extract and SUBSTRING builtin functions

### PR DESCRIPTION
**Pull request by nvb**
See #2042

Note that the substring functions do not support `substring(string for int)`. `substring(string for int from int)` is also not supported (yet?) because it is not valid SQL99.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3049)

### CODE REVIEW

File: sql/parser/sql.y [hunk 1]
```diff
@@ -3094,18 +3098,36 @@ position_list:
...
 | a_expr substr_for { unimplemented() }
-| expr_list { unimplemented() }
-| /* EMPTY */ {}
+| expr_list 
+  { 
+    $$ = $1
+  }
+| /* EMPTY */ 
+  {
+    $$ = nil
```
Comment (tamird) [inline sql/parser/sql.y]: am I reading this right that `SUBSTRING()` parses? is that just so the error gets generated at dispatch?

**Comment by tamird** [comment]
could you add some parser tests for the functionality you've introduced?

File: sql/parser/sql.y (hunk 1 above)
Comment (nvanbenschoten) [inline sql/parser/sql.y]: Yeah, that will allow the error "unknown signature for SUBSTRING: SUBSTRING()" to be generated later. Added a test for this.

**Comment by petermattis** [comment]
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/parser/sql.y, line 3105 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gmYhdRGVsYETNQV_A)** ([raw file](https://github.com/cockroachdb/cockroach/blob/77a53267ad432e0368c7cd1a3efb8bbd83c5f412/sql/parser/sql.y#L3105)):</sup>
Is `SUBSTRING(a FOR 1 FROM 2)` different from `SUBSTRING(a FROM 1 FOR 2)`?


---

<sup>**[sql/parser/sql.y, line 3110 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gn2P_ifLRQ-zMIXvj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/77a53267ad432e0368c7cd1a3efb8bbd83c5f412/sql/parser/sql.y#L3110)):</sup>
Looks like this is `SUBSTRING($1, 0, $2)`.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Comment by nvanbenschoten** [comment]
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[sql/parser/sql.y, line 3105 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gmYhdRGVsYETNQV_A)** ([raw file](https://github.com/cockroachdb/cockroach/blob/77a53267ad432e0368c7cd1a3efb8bbd83c5f412/sql/parser/sql.y#L3105)):</sup>
`SUBSTRING(a FOR 1 FROM 2)` should be equivalent to `SUBSTRING(a FROM 2 FOR 1)`

According to [gram.y](http://sourceforge.net/p/postgres-xl/postgres-xl/ci/eef28c9873e709e6d8c4f625290fc12f48935dab/tree/src/backend/parser/gram.y), it's not technically valid SQL99, but we might as well support it. Done.


---

<sup>**[sql/parser/sql.y, line 3110 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gn2P_ifLRQ-zMIXvj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/77a53267ad432e0368c7cd1a3efb8bbd83c5f412/sql/parser/sql.y#L3110)):</sup>
Yep, just added support for it. I had forgotten that I could just wrap a constant in a `DInt` to get an `Expr`.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Comment by tamird** [comment]
Reviewed 1 of 2 files at r1, 2 of 5 files at r2, 1 of 1 files at r3, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/parser/parse_test.go, line 486 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gvcEiicVUxboGLO6l)** ([raw file](https://github.com/cockroachdb/cockroach/blob/144ba981435c6857be32fb6450c1f0ad426f17b3/sql/parser/parse_test.go#L486)):</sup>
should add some cases for the rearranged variant you added


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Comment by nvanbenschoten** [comment]
Review status: 2 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[sql/parser/parse_test.go, line 486 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gvcEiicVUxboGLO6l)** ([raw file](https://github.com/cockroachdb/cockroach/blob/144ba981435c6857be32fb6450c1f0ad426f17b3/sql/parser/parse_test.go#L486)):</sup>
I'm glad you suggested that, because it made me realize that SUBSTRING's "from" argument was supposed to be 1-indexed. Added tests and made a change to the default "from" argument.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Comment by tamird** [comment]
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Comment by petermattis** [comment]
LGTM


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Comment by nvanbenschoten** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[sql/parser/sql.y, line 3110 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gn2P_ifLRQ-zMIXvj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/77a53267ad432e0368c7cd1a3efb8bbd83c5f412/sql/parser/sql.y#L3110)):</sup>
How do I resolve these issues when the CL looks good? Does it matter if they're left unresolved?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[sql/parser/sql.y, line 3110 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gn2P_ifLRQ-zMIXvj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/77a53267ad432e0368c7cd1a3efb8bbd83c5f412/sql/parser/sql.y#L3110)):</sup>
There's a checkmark button you should be able to click to mark the discussion resolved.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Comment by nvanbenschoten** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[sql/parser/sql.y, line 3110 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gn2P_ifLRQ-zMIXvj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/77a53267ad432e0368c7cd1a3efb8bbd83c5f412/sql/parser/sql.y#L3110)):</sup>
Strange, the only checkmark I'm seeing doesn't seem to be related to resolving the discussion, only to the current message's response requirements. It still says the overall discussion is waiting for a response from Peter.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Comment by petermattis** [comment]
Reviewed 1 of 5 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[sql/parser/sql.y, line 3110 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049#-K2gn2P_ifLRQ-zMIXvj)** ([raw file](https://github.com/cockroachdb/cockroach/blob/77a53267ad432e0368c7cd1a3efb8bbd83c5f412/sql/parser/sql.y#L3110)):</sup>
Heh, don't get hung-up on all of the reviewable.io checks. 


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3049)

**Pull request merged**